### PR TITLE
perf: Make date_range / datetime_range ~10x faster for constant durations

### LINKED ISSUE #19179: Performance issues when using `pl.datetime_range` (compared to pandas)
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import datetime as dt
import polars as pl
import pandas as pd

start = dt.date(2017, 1, 1)
end = dt.datetime(2022, 12, 31, 23, 0)

%time pd.date_range(start, end, freq='H')
%time pl.datetime_range(start, end, interval=dt.timedelta(seconds=3600), eager=True)
```

### Log output

```shell
# pandas:
CPU times: user 181 µs, sys: 9 µs, total: 190 µs
Wall time: 193 µs

# polars:
CPU times: user 551 µs, sys: 301 µs, total: 852 µs
Wall time: 443 µs
```

### Issue description

Mostly curious if there's a reason as to why there's such a big difference in performance. Note of course the time is variable in both pandas and polars, but I noticed pandas tends to stay between 0.2-0.6ms, whereas polars can swing between 0.8 - >2ms.

### Expected behavior

Obviously, I want polars to be as fast (or even faster) than pandas (:

### Installed versions

<details>

```
--------Version info---------
Polars:              1.9.0
Index type:          UInt32
Platform:            macOS-14.2-arm64-arm-64bit
Python:              3.11.9 (main, Apr  2 2024, 08:25:04) [Clang 15.0.0 (clang-1500.3.9.4)]
----Optional dependencies----
adbc_driver_manager  <not installed>
altair               <not installed>
cloudpickle          3.0.0
connectorx           <not installed>
deltalake            <not installed>
fastexcel            0.10.4
fsspec               2024.3.1
gevent               <not installed>
great_tables         <not installed>
matplotlib           3.8.4
nest_asyncio         1.6.0
numpy                1.26.4
openpyxl             3.1.5
pandas               1.5.3
pyarrow              16.0.0
pydantic             2.6.3
pyiceberg            <not installed>
sqlalchemy           2.0.32
torch                <not installed>
xlsx2csv             <not installed>
xlsxwriter           <not installed>
```

</details>

### PR DESCRIPTION
**Pull request by MarcoGorelli**
closes #19179 

from https://www.kaggle.com/code/marcogorelli/polars-timing/notebook?scriptVersionId=200783587 it's ~10x faster

EDIT: disregard the name of the branch 🙈

### CODE REVIEW

File: py-polars/tests/unit/functions/range/test_datetime_range.py [hunk 1]
```diff
@@ -579,3 +581,34 @@ def test_datetime_range_specifying_ambiguous_11713() -> None:
...
+
+
+@given(
+    closed=st.sampled_from(["none", "left", "right", "both"]),
+    time_unit=st.sampled_from(["ms", "us", "ns"]),
+    n=st.integers(1, 100),
+    unit=st.sampled_from(["s", "m", "h", "d", "mo"]),
+    start=st.datetimes(datetime(1965, 1, 1), datetime(2100, 1, 1)),
+)
+def test_datetime_range_fast_slow_paths(
```
Comment (MarcoGorelli) [inline py-polars/tests/unit/functions/range/test_datetime_range.py:593]: this runs in about 0.71s. well-worth it imo

Comment (ritchie46) [reply py-polars/tests/unit/functions/range/test_datetime_range.py:593]: Can we only run this in the benchmark test? That means they run on a perf release. I do think 0.71s is very long ;)

File: crates/polars-time/src/date_range.rs [hunk 2]
```diff
@@ -111,25 +111,43 @@ pub(crate) fn datetime_range_i64(
...
+    };
+    if interval.is_constant_duration(time_zone_opt_string.as_deref()) {
+        // Fast path!
+        return match closed {
+            ClosedWindow::Both => Ok((start..=end)
+                .step_by(duration as usize)
+                .collect::<Vec<i64>>()),
+            ClosedWindow::None => Ok((start..end)
+                .step_by(duration as usize)
+                .skip(1)
```
Comment (ritchie46) [inline crates/polars-time/src/date_range.rs:132]: Nit: can we do this skip in the `(start..end)` range directly? That might produce faster code.

File: crates/polars-time/src/date_range.rs [hunk 3]
```diff
@@ -111,25 +111,43 @@ pub(crate) fn datetime_range_i64(
...
+    let time_zone_opt_string: Option<String> = match tz {
+        #[cfg(feature = "timezones")]
+        Some(tz) => Some(tz.to_string()),
+        _ => None,
+    };
+    if interval.is_constant_duration(time_zone_opt_string.as_deref()) {
+        // Fast path!
+        return match closed {
+            ClosedWindow::Both => Ok((start..=end)
+                .step_by(duration as usize)
```
Comment (edwinvehmaanpera) [inline crates/polars-time/src/date_range.rs:128]: I was playing around trying to improve the performance as a learning exercise and ended up basically with the same solution. But one thing that gave me pause was would casting i64 to usize here potentially break with very large durations and 32 bit systems (does Polars even support 32bit)?

Comment (ritchie46) [reply crates/polars-time/src/date_range.rs:128]: > does Polars even support 32bit

Not really. People could compile it on 32 bits, but it's not something we want to focus on. We can replace with `try_into`.

File: crates/polars-time/src/date_range.rs (hunk 2 above)
Comment (MarcoGorelli) [reply crates/polars-time/src/date_range.rs]: I couldn't find a way to do this, suggestions I looked for how to exclude the first element resulted in `skip(1)` - do you have a suggestion?

Comment (edwinvehmaanpera) [reply crates/polars-time/src/date_range.rs]: The way I did it was `(start+step..end)`.

Comment (MarcoGorelli) [reply crates/polars-time/src/date_range.rs]: right, indeed 🙏

**Review by ritchie46 — APPROVED** [review]
Thank you @MarcoGorelli. Looks great now. Afaic it can be undrafted. :)

**Comment by MarcoGorelli** [comment]
thanks - i tried replacing the hypothesis test with a cartesian product over all inputs, but then it took 12 seconds 😱 

like this it's ~0.33s, and marked with benchmark

**Comment by ritchie46** [comment]
> like this it's ~0.33s, and marked with benchmark

Still terrible if we have to do that for our whole API. :') I think we shuold accept we sample all arguments on multiple CI runs, not per-se one. :sweat_smile:

**Comment by d-reynol** [comment]
@MarcoGorelli will this improve the performance of date_range**s** as well?

It looks like it should, but I'm by no means an expert

**Comment by MarcoGorelli** [comment]
yup!

**Comment by d-reynol** [comment]
ah, of course there's a catch :) - I'm using months.

Thank you!

**Pull request merged**